File: Data Structures & Algorithms/jump-game/submission-0.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)

        i = 1
        jump = nums[0]

        while i < n:
            if i <= jump:
                jump = max(jump, i + nums[i])
            else:
                return False
            i+=1

        return True 
            

File: Data Structures & Algorithms/burst-balloons/submission-0.cpp
class Solution {
public:
    vector<vector<int>> dp;
    int burst(vector<int>&nums, int l, int r) {

        int ans = 0;

        if (dp[l][r] != -1) return dp[l][r];
        
        for (int i = l+1; i < r; i++) {
            int temp = nums[l]*nums[i]*nums[r] + burst(nums, l, i) + burst(nums, i, r);
            ans = max(ans, temp);
        }
        
        return dp[l][r] = ans;    
    }
    int maxCoins(vector<int>& nums) {
        
        nums.insert(nums.begin(), 1);
        int n = nums.size();
        nums.push_back(1);
        dp.resize(n+2, vector<int>(n+2, -1));
        return burst(nums, 0, nums.size()-1);
    }
};
